<!DOCTYPE html>
<html lang="es">
<head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Manejo de Excepciones</title>

    <link rel="alternate" href="https://campusempresa.com/mod/haskell/06-04-exception-handling" hreflang="es" />
	<link rel="alternate" href="https://campusempresa.cat/mod/haskell/06-04-exception-handling" hreflang="ca" />
	<link rel="alternate" href="https://enterprisecampus.net/mod/haskell/06-04-exception-handling" hreflang="en" />
    
	<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.3/dist/css/bootstrap.min.css" rel="stylesheet">
	<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap-icons/font/bootstrap-icons.css">
	<link href="/css/site.css" rel="stylesheet">
	
    <script src="https://code.jquery.com/jquery-3.7.1.min.js"></script>
  	<script type="text/javascript" src="/js/math_init.js"></script>
  	<script type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/startup.js"></script>
  	<script type="text/javascript" src="/js/cookie.js"></script>
  	<script type="text/javascript" src="/js/main.js"></script>
</head>

<body >
    <div id="content">
		<div id="header" class="container-xxl">
	<div class="row">
		<div class="col-12 col-md-6 p-2 p-md-0">
			<h1 class="m-0 p-0">
				<a href="/"><img src="/img/logo_header.png"></a>
			</h1>
		</div>
		<div class="col-12 col-md-6 p-2 p-md-0 text-end">
				<a href="https://enterprisecampus.net/mod/haskell/06-04-exception-handling" class="px-2">EN</a></b>
	|
	<b class="px-2">ES</b>
	|
	<a href="https://campusempresa.cat/mod/haskell/06-04-exception-handling" class="px-2">CA</a>
<br>
			<cite>Construyendo la sociedad de hoy y del mañana</cite>
		</div>
	</div>
</div>
<div id="subheader" class="container-xxl">
	<div class="row">
		<div class="col-12 p-2 p-md-0 m-0 text-end">
			<a href="/objective">El Proyecto</a> | 
<a href="/about">Sobre nosotros</a> | 
<a href="/contribute">Contribuir</a> | 
<a href="/donate">Donaciones</a> | 
<a href="/licence">Licencia</a>
		</div>
	</div>
</div>

<div class="top-bar container-fluid">
	<div class="container-xxl">
		<div class="row">
			<div class="col" id="left_menu">
									<a href="./">Contenido del curso</a>
					<span class="sep">|</span>
								<a href="/all/competencias">Competencias técnicas</a>
				<a href="/all/conocimientos">Conocimientos</a>
				<a href="/all/soft_skills">Competencias sociales</a>
			</div>
		</div>
	</div>
</div>

<div class="container-xxl" id="main_content">
	<div class="row">
		<div class="col-12 col-lg-8">
			<div id="nav1" class="navigation"></div>
			<div id="inner_content">
								<div class='row navigation'>
	<div class='col-1 col-md-2'>
					<a href='06-03-interacting-with-system' title="Interacción con el Sistema">
				<span class="d-none d-md-inline">&#x25C4; Anterior</span>
				<span class="d-inline d-md-none"><i class="bi bi-caret-left-square-fill"></i></span>
			</a>
			</div>
	<div class='col-10 col-md-8 text-center'>
					<a href="./"><h2 style="text-decoration:underline">Manejo de Excepciones</h2></a>
			</div>
	<div class='col-1 col-md-2 text-end'>
					<a href='07-01-concurrency' title="Concurrencia">
				<span class="d-none d-md-inline">Siguiente &#x25BA;</span>
				<span class="d-inline d-md-none"><i class="bi bi-caret-right-square-fill"></i></span>
			</a>
			</div>
</div>
<div class='content'><p>En Haskell, el manejo de excepciones es un aspecto crucial para escribir programas robustos y confiables. A diferencia de muchos lenguajes imperativos, Haskell maneja las excepciones de una manera que se integra con su naturaleza funcional. En este tema, aprenderemos cómo capturar y manejar excepciones en Haskell.</p>
</div><h1><p>Conceptos Clave</p>
</h1>
<div class='content'><ol>
<li><strong>Excepciones en Haskell</strong>: Haskell proporciona un sistema de excepciones que permite a los programas manejar errores de manera controlada.</li>
<li><strong>Módulo <code>Control.Exception</code></strong>: Este módulo contiene las funciones y tipos necesarios para trabajar con excepciones.</li>
<li><strong>Tipos de Excepciones</strong>: Haskell distingue entre excepciones síncronas (errores de programación) y asíncronas (interrupciones externas).</li>
</ol>
</div><h1><p>Módulo <code>Control.Exception</code></p>
</h1>
<div class='content'><p>El módulo <code>Control.Exception</code> es la base para el manejo de excepciones en Haskell. Aquí están algunas de las funciones y tipos más importantes:</p>
<ul>
<li><code>catch</code>: Captura una excepción y permite manejarla.</li>
<li><code>try</code>: Similar a <code>catch</code>, pero devuelve un <code>Either</code> en lugar de lanzar una excepción.</li>
<li><code>throwIO</code>: Lanza una excepción en el contexto de IO.</li>
<li><code>SomeException</code>: Tipo de datos que puede representar cualquier excepción.</li>
</ul>
</div><h2><p>Ejemplo Básico de <code>catch</code></p>
</h2>
<div class='content'></div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("aW1wb3J0IENvbnRyb2wuRXhjZXB0aW9uCgptYWluIDo6IElPICgpCm1haW4gPSBkbwogICAgcmVzdWx0IDwtIHRyeSAocmVhZEZpbGUgIm5vbmV4aXN0ZW50LnR4dCIpIDo6IElPIChFaXRoZXIgSU9FeGNlcHRpb24gU3RyaW5nKQogICAgY2FzZSByZXN1bHQgb2YKICAgICAgICBMZWZ0IGV4ICAtPiBwdXRTdHJMbiAkICJDYXVnaHQgZXhjZXB0aW9uOiAiICsrIHNob3cgZXgKICAgICAgICBSaWdodCBjb250ZW50IC0+IHB1dFN0ckxuIGNvbnRlbnQ="))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>import Control.Exception

main :: IO ()
main = do
    result &lt;- try (readFile &quot;nonexistent.txt&quot;) :: IO (Either IOException String)
    case result of
        Left ex  -&gt; putStrLn $ &quot;Caught exception: &quot; ++ show ex
        Right content -&gt; putStrLn content</pre></div><div class='content'><p>En este ejemplo, intentamos leer un archivo que no existe. La función <code>try</code> captura la excepción y devuelve un <code>Either</code> que puede ser manejado de manera segura.</p>
</div><h2><p>Ejemplo de <code>throwIO</code></p>
</h2>
<div class='content'></div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("aW1wb3J0IENvbnRyb2wuRXhjZXB0aW9uCgptYWluIDo6IElPICgpCm1haW4gPSBkbwogICAgdGhyb3dJTyAodXNlckVycm9yICJUaGlzIGlzIGEgdXNlciBlcnJvciIpCiAgICBwdXRTdHJMbiAiVGhpcyB3aWxsIG5vdCBiZSBwcmludGVkIg=="))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>import Control.Exception

main :: IO ()
main = do
    throwIO (userError &quot;This is a user error&quot;)
    putStrLn &quot;This will not be printed&quot;</pre></div><div class='content'><p>En este ejemplo, lanzamos una excepción usando <code>throwIO</code>. La línea siguiente no se ejecutará debido a la excepción.</p>
</div><h1><p>Tipos de Excepciones</p>
</h1>
<div class='content'><p>Haskell tiene varios tipos de excepciones, pero las más comunes son:</p>
<ul>
<li><strong>IOException</strong>: Errores relacionados con la entrada/salida.</li>
<li><strong>ArithException</strong>: Errores aritméticos como división por cero.</li>
<li><strong>AsyncException</strong>: Excepciones asíncronas como interrupciones de usuario.</li>
</ul>
</div><h2><p>Manejo de Excepciones Específicas</p>
</h2>
<div class='content'><p>Podemos capturar excepciones específicas utilizando patrones de coincidencia:</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("aW1wb3J0IENvbnRyb2wuRXhjZXB0aW9uCgptYWluIDo6IElPICgpCm1haW4gPSBkbwogICAgcmVzdWx0IDwtIHRyeSAoZXZhbHVhdGUgKDUgYGRpdmAgMCkpIDo6IElPIChFaXRoZXIgQXJpdGhFeGNlcHRpb24gSW50KQogICAgY2FzZSByZXN1bHQgb2YKICAgICAgICBMZWZ0IGV4ICAtPiBwdXRTdHJMbiAkICJDYXVnaHQgYXJpdGhtZXRpYyBleGNlcHRpb246ICIgKysgc2hvdyBleAogICAgICAgIFJpZ2h0IHZhbCAtPiBwdXRTdHJMbiAkICJSZXN1bHQ6ICIgKysgc2hvdyB2YWw="))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>import Control.Exception

main :: IO ()
main = do
    result &lt;- try (evaluate (5 `div` 0)) :: IO (Either ArithException Int)
    case result of
        Left ex  -&gt; putStrLn $ &quot;Caught arithmetic exception: &quot; ++ show ex
        Right val -&gt; putStrLn $ &quot;Result: &quot; ++ show val</pre></div><div class='content'><p>En este ejemplo, capturamos una excepción aritmética específica (división por cero).</p>
</div><h1><p>Ejercicios Prácticos</p>
</h1>
<div class='content'></div><h2><p>Ejercicio 1: Captura de Excepciones de IO</p>
</h2>
<div class='content'><p>Escribe un programa que intente leer un archivo y capture cualquier excepción de IO, imprimiendo un mensaje de error adecuado.</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("aW1wb3J0IENvbnRyb2wuRXhjZXB0aW9uCmltcG9ydCBTeXN0ZW0uSU8KCm1haW4gOjogSU8gKCkKbWFpbiA9IGRvCiAgICBwdXRTdHJMbiAiRW50ZXIgdGhlIGZpbGVuYW1lOiIKICAgIGZpbGVuYW1lIDwtIGdldExpbmUKICAgIHJlc3VsdCA8LSB0cnkgKHJlYWRGaWxlIGZpbGVuYW1lKSA6OiBJTyAoRWl0aGVyIElPRXhjZXB0aW9uIFN0cmluZykKICAgIGNhc2UgcmVzdWx0IG9mCiAgICAgICAgTGVmdCBleCAgLT4gcHV0U3RyTG4gJCAiRXJyb3IgcmVhZGluZyBmaWxlOiAiICsrIHNob3cgZXgKICAgICAgICBSaWdodCBjb250ZW50IC0+IHB1dFN0ckxuIGNvbnRlbnQ="))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>import Control.Exception
import System.IO

main :: IO ()
main = do
    putStrLn &quot;Enter the filename:&quot;
    filename &lt;- getLine
    result &lt;- try (readFile filename) :: IO (Either IOException String)
    case result of
        Left ex  -&gt; putStrLn $ &quot;Error reading file: &quot; ++ show ex
        Right content -&gt; putStrLn content</pre></div><div class='content'></div><h2><p>Ejercicio 2: Manejo de Excepciones Aritméticas</p>
</h2>
<div class='content'><p>Escribe un programa que realice una operación aritmética y capture cualquier excepción aritmética, imprimiendo un mensaje de error adecuado.</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("aW1wb3J0IENvbnRyb2wuRXhjZXB0aW9uCgptYWluIDo6IElPICgpCm1haW4gPSBkbwogICAgcHV0U3RyTG4gIkVudGVyIHR3byBudW1iZXJzIHRvIGRpdmlkZToiCiAgICBudW0xIDwtIHJlYWRMbgogICAgbnVtMiA8LSByZWFkTG4KICAgIHJlc3VsdCA8LSB0cnkgKGV2YWx1YXRlIChudW0xIGBkaXZgIG51bTIpKSA6OiBJTyAoRWl0aGVyIEFyaXRoRXhjZXB0aW9uIEludCkKICAgIGNhc2UgcmVzdWx0IG9mCiAgICAgICAgTGVmdCBleCAgLT4gcHV0U3RyTG4gJCAiQXJpdGhtZXRpYyBlcnJvcjogIiArKyBzaG93IGV4CiAgICAgICAgUmlnaHQgdmFsIC0+IHB1dFN0ckxuICQgIlJlc3VsdDogIiArKyBzaG93IHZhbA=="))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>import Control.Exception

main :: IO ()
main = do
    putStrLn &quot;Enter two numbers to divide:&quot;
    num1 &lt;- readLn
    num2 &lt;- readLn
    result &lt;- try (evaluate (num1 `div` num2)) :: IO (Either ArithException Int)
    case result of
        Left ex  -&gt; putStrLn $ &quot;Arithmetic error: &quot; ++ show ex
        Right val -&gt; putStrLn $ &quot;Result: &quot; ++ show val</pre></div><div class='content'></div><h1><p>Resumen</p>
</h1>
<div class='content'><p>En esta sección, hemos aprendido cómo manejar excepciones en Haskell utilizando el módulo <code>Control.Exception</code>. Hemos visto cómo capturar excepciones específicas y cómo lanzar nuestras propias excepciones. Estos conceptos son fundamentales para escribir programas Haskell robustos y confiables.</p>
<p>En el siguiente módulo, exploraremos temas avanzados como la concurrencia y el paralelismo, que nos permitirán escribir programas más eficientes y escalables.</p>
</div><div class='row navigation'>
	<div class='col-1 col-md-2'>
					<a href='06-03-interacting-with-system' title="Interacción con el Sistema">
				<span class="d-none d-md-inline">&#x25C4; Anterior</span>
				<span class="d-inline d-md-none"><i class="bi bi-caret-left-square-fill"></i></span>
			</a>
			</div>
	<div class='col-10 col-md-8 text-center'>
			</div>
	<div class='col-1 col-md-2 text-end'>
					<a href='07-01-concurrency' title="Concurrencia">
				<span class="d-none d-md-inline">Siguiente &#x25BA;</span>
				<span class="d-inline d-md-none"><i class="bi bi-caret-right-square-fill"></i></span>
			</a>
			</div>
</div>

			</div>
		</div>
		<div class="col-12 col-lg-4 publi" id="div_publi">
			<h1>Publicidad</h1>
			<p>Este espacio está destinado a publicidad.</p>
			<p>Si quieres ser patrocinador, contáctanos para incluir enlaces en esta zona: <a href='mailto:admin@campusempresa.cat'>admin@campusempresa.cat</a></p>
			<p>¡Gracias por colaborar!</p>
		</div>
	</div>
</div>

   <div class="container-xxl my-3">
	<div class="row">
		<div class="col">
			<footer>&copy; Copyright 2024. Todos los derechos reservados</footer>
		</div>
	</div>
</div>	

<div id="cookies_adv" style="display:none;">
	Usamos cookies para mejorar tu experiencia de uso y ofrecer contenidos adaptados a tus intereses.
    <a href="#" id="btn_accept_cookies" class="button">Aceptar</a>
    <a href="/cookies">Mas información</a>
</div>	

	</div>    
</body>
</html>
